### DISCUSSION TITLE
Multi-node training using sharding API

### DISCUSSION THREAD

**Discussion by neel04** [jax-ml/jax#20053] [Q&A]
I'm using TPUs, and wished to do training across multi-node/hosts. The official docs suggest using `xmap`/`pmap`.

However, I'm using the sharding API to shard across multiple local devices.

So how can we extend the sharding to accomodate a multi-node setup?

AIUI, we should be able to to provide a sort of 3D sharding like (2, 8, 1) for 2x TPUs with 8 local devices each, DDP styled. 
This would allow us to switch between n-way data parallelism and m-way model parallelism as outlined [here](https://jax.readthedocs.io/en/latest/notebooks/Distributed_arrays_and_automatic_parallelization.html#examples-neural-networks).

But this doesn't seem to be the case?

Related: SO bounty thread I started [here](https://stackoverflow.com/questions/78075943/multi-node-host-training-with-the-sharding-api).

**Comment by wen020** [comment]
Do you currently know the best practices for implementing multi-host data parallelism with JAX?

**Reply by neel04** [reply]
I ended up using [scalax](https://github.com/young-geng/scalax)

**Reply by wen020** [reply]
Can Scalax be used for distributed training on multiple machines?

**Reply by neel04** [reply]
yes

**Comment by ayaka14732** [comment]
> However, I'm using the sharding API to shard across multiple local devices. So how can we extend the sharding to accomodate a multi-node setup?

Just change the code from:

```python
arr = jax.device_put(arr, NamedSharding(mesh, P(*partition_spec)))
```

To:

```python
arr = jax.make_array_from_callback(arr.shape, NamedSharding(mesh, P(*partition_spec)), lambda idx: arr[idx]) 
```

See https://github.com/google/jax/discussions/20041

**Reply by neel04** [reply]
is `idx` the `host_id`?

**Comment by yashk2810** [comment]
If your input pipeline is fully data parallel, take a look at the docstring of: https://jax.readthedocs.io/en/latest/_autosummary/jax.make_array_from_single_device_arrays.html

If am also planning to expose a helper function to do just this! Are you also asking on how to shard the weights?

**Reply by neel04** [reply]
Yes please. I'd prefer a complete jax example for TP, DDP and FSDP 🙏

**Reply by sourcesync** [reply]
Hey @neel04 Did you ever get this to work on multiple machines?
